hibikase/PlaybackBarWidget.cpp

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+// This file implements a slider control that represents the current playback
+// position. Hibikase previously used QSlider for this, but its behaviour is not
+// consistent across platforms, and it can be quite glitchy. Using a custom
+// slider also lets us show what parts of the song have timed lines.
+
+#include <algorithm>
+#include <chrono>
+
+#include <QBrush>
+#include <QColor>
+#include <QPainter>
+#include <QPaintEvent>
+#include <QPen>
+#include <QPoint>
+#include <QRect>
+
+#include "PlaybackBarWidget.h"
+
+using Microseconds = std::chrono::microseconds;
+
+const int HEIGHT = 20;
+const int LINE_HEIGHT = 4;
+const int HANDLE_WIDTH = 8;
+
+PlaybackBarWidget::PlaybackBarWidget(QWidget *parent)
+    : QWidget(parent), m_current_time{}, m_song_length{}, m_being_dragged{},
+      m_song_ref{}
+{
+    setMinimumHeight(HEIGHT);
+    setVisible(true);
+    setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::Fixed);
+}
+
+Microseconds PlaybackBarWidget::GetCurrentTime() const
+{
+    return m_current_time;
+}
+
+bool PlaybackBarWidget::IsBeingDragged() const
+{
+    return m_being_dragged;
+}
+
+void PlaybackBarWidget::Update(Microseconds current_time, Microseconds song_length)
+{
+    m_current_time = current_time;
+    m_song_length = song_length;
+    update();
+}
+
+QRect PlaybackBarWidget::ComputeLineRect() const
+{
+    int w = width() - HANDLE_WIDTH;
+    int h = LINE_HEIGHT;
+
+    return QRect { (width() - w) / 2, (HEIGHT - h) / 2, w, h };
+}
+QRectF PlaybackBarWidget::ComputeHandleRect() const
+{
+    QRect line_rect = ComputeLineRect();
+    qreal time = (double(m_current_time.count()) / m_song_length.count());
+    return QRectF { line_rect.width() * time, 0, qreal(HANDLE_WIDTH), qreal(HEIGHT) };
+}
+
+void PlaybackBarWidget::paintEvent(QPaintEvent* e)
+{
+    if (m_song_length.count() == 0)
+    {
+        return;
+    }
+
+    QPainter painter(this);
+    painter.setRenderHint(QPainter::Antialiasing);
+
+    auto line_color = QColor(~QPalette().color(QPalette::Text).rgb());
```

Comment:
Preferably:

```suggestion
    auto line_color = QColor(~(QPalette().color(QPalette::Text).rgb()));
```

So that it looks less like this has something to do with the destructor `~QPalette`.